### PR TITLE
Detect types stubs dependencies

### LINKED ISSUE #221: Detect dependencies used for types
**Is your feature request related to a problem? Please describe.**

My project uses:
* `requests` (as an example)
*  `mypy` for type checking
* and therefore [`types-requests`](https://pypi.org/project/types-requests/) from [typeshed](https://github.com/python/typeshed/).

FawltyDeps reports `types-requests` as unused, together with other typeshed packages.

**Describe the solution you'd like**

I'd like FawltyDeps to consider `types-ABCD` used if and only if `ABCD` is used (or re-use the logic from `mypy` in case there are non-trivial rules for which packages declare stubs for which).

This can be behind a flag (e.g. `typeshed = true`) in configuration.

**Describe alternatives you've considered**

I can manually ignore the `types-*` packages using `ignore_unused`, but then if `requests` is removed `types-requests` will need to be removed manually, and if one forgets that it'll become fawlty (tm?) - the problem this project is trying to prevent :-)

### PR DESCRIPTION
**Pull request by zz1874**
Closes #221 

This PR adds a functionality that detects types stubs dependencies (from typeshed).

Right now, the `self_test` CI didn't pass, so I'll keep this PR as draft PR and will take a look tommorrow.

### CODE REVIEW

**Comment by Nour-Mws** [comment]
The self-test runs correctly outside Nox (using the Poetry venv).
I extracted the JSON of the self-test command run with the Nox venv as follows:
```
fawltydeps --pyenv .nox/self_test-3-11 --json | jq .resolved_deps | grep "types-setuptools" -A 7
```
Apparently, `types-setuptools` is being resolved by IdentityMapping in the Nox env:
```
  "types-setuptools": {
    "package_name": "types_setuptools",
    "import_names": [
      "types_setuptools"
    ],
    "resolved_with": "IdentityMapping",
    "debug_info": null
  },

```

File: fawltydeps/packages.py [hunk 1]
```diff
@@ -83,9 +83,26 @@ def __post_init__(self) -> None:
...
+            provides_stubs_for = [
+                import_name.removesuffix("-stubs") for import_name in self.import_names
+            ]
+        else:
+            provides_stubs_for = [
+                import_name[: -len("-stubs")]
+                for import_name in self.import_names
+                if import_name.endswith("-stubs")
+            ]
+        return set(provides_stubs_for)
```
[the remark this answers is not in the document]
Comment (zz1874) [reply fawltydeps/packages.py]: Yes that makes good sense.
Addressed in https://github.com/tweag/FawltyDeps/pull/362/commits/e99c61b0708f0495e95972246a8f894426054068

**Comment by zz1874** [comment]
This PR is ready for review @Nour-Mws 
It has many commits, and they are only for experimenting with caches. Feel free to ignore them :smile: 
The CI tests didn't pass in the end, and I opened another PR (#363) to fix it.

Also, please note that I had to re-generate the caches in GitHub Actions CI for multiple times, as soon as this PR and the PR for fixing the tests are improved, I'll delete the duplicate caches manually.

**Comment by Nour-Mws** [comment]
This looks good. As soon as PR #363 is merged we can rebase this and re-run the CI.

**Review by Nour-Mws — APPROVED** [review]
Good to go!

**Pull request merged**